File: das/retrieval.py
"""Two-stage retrieval, hand-built (same shape as personal-rag).

Stage 1 — candidates: semantic top-8 (cosine over MiniLM embeddings)
                      ∪ lexical top-8 (hand-built BM25)
Stage 2 — rerank:     cross-encoder scores every candidate against the query,
                      keep top 5.

Results carry everything a reader needs to verify: excerpt, program, date,
timecode, ASR confidence, tags (marked AI-generated) and the archive.org
source link. No index entry, no answer — results only ever come from the corpus.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

from das.bm25 import BM25

ROOT = Path(__file__).resolve().parents[1]
RECORD_DIR = ROOT / "data" / "records"
INDEX_DIR = ROOT / "data" / "index"

EMBED_MODEL = "all-MiniLM-L6-v2"
RERANK_MODEL = "cross-encoder/ms-marco-MiniLM-L-6-v2"
CANDIDATES_PER_STAGE = 8
TOP_K = 5


class Searcher:
    def __init__(self) -> None:
        import numpy as np
        from sentence_transformers import CrossEncoder, SentenceTransformer

        self.windows = json.loads((INDEX_DIR / "windows.json").read_text(encoding="utf-8"))
        self.embeddings = np.load(INDEX_DIR / "embeddings.npy")
        self.records = {
            r["id"]: r
            for r in (
                json.loads(p.read_text(encoding="utf-8"))
                for p in RECORD_DIR.glob("*.json")
            )
        }
        self.bm25 = BM25([w["text"] for w in self.windows])
        self.embedder = SentenceTransformer(EMBED_MODEL)
        self.reranker = CrossEncoder(RERANK_MODEL)

    def search(self, query: str, k: int = TOP_K) -> list[dict]:
        import numpy as np

        # Stage 1: candidate union
        qvec = self.embedder.encode([query], normalize_embeddings=True)[0]
        sims = self.embeddings @ qvec
        semantic = set(np.argsort(-sims)[:CANDIDATES_PER_STAGE].tolist())
        lexical = {i for i, _ in self.bm25.top(query, CANDIDATES_PER_STAGE)}
        candidates = sorted(semantic | lexical)
        if not candidates:
            return []

        # Stage 2: rerank
        pairs = [(query, self.windows[i]["text"]) for i in candidates]
        scores = self.reranker.predict(pairs)
        ranked = sorted(zip(candidates, scores), key=lambda x: x[1], reverse=True)[:k]

        results = []
        for idx, score in ranked:
            w = self.windows[idx]
            rec = self.records[w["clip_id"]]
            results.append({
                "clip_id": w["clip_id"],
                "title": rec["title"],
                "program": rec["program"],
                "date": rec["date"],
                "decade": rec["decade"],
                "excerpt": w["text"],
                "start_s": w["start"],
                "end_s": w["end"],
                "asr_confidence": w["confidence"],
                "rerank_score": round(float(score), 4),
                "matched_by": ("semantic" if idx in semantic else "")
                              + ("+lexical" if idx in lexical and idx in semantic else "")
                              + ("lexical" if idx in lexical and idx not in semantic else ""),
                "tags": rec["tags"],
                "source_url": rec["source_url"],
                "rights": rec["rights"],
                "audio_file": rec["audio_file"],
            })
        return results


def main() -> int:
    query = " ".join(sys.argv[1:]) or "allied invasion of normandy"
    searcher = Searcher()
    print(f'Query: "{query}"\n')
    for i, r in enumerate(searcher.search(query), 1):
        mins, secs = divmod(int(r["start_s"]), 60)
        print(f"{i}. [{r['rerank_score']:+.2f}] {r['program']} {r['date']} @ {mins}:{secs:02d}  ({r['matched_by']})")
        print(f"   {r['title']}  (asr_conf={r['asr_confidence']:.2f})")
        print(f"   \"{r['excerpt'][:160]}\"")
        print(f"   {r['source_url']}\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: demo/site/retrieval.mjs
// Hand-built two-stage retrieval — a line-for-line port of das/bm25.py and
// das/retrieval.py to JavaScript, so the in-browser demo runs the SAME
// pipeline as the Python app: semantic top-8 ∪ BM25 top-8 → cross-encoder
// rerank → top-5, best window per clip.
//
// This module is pure JS (no model code). The embedder and reranker are
// injected, so the same file runs in the browser demo and in the Node
// eval harness (demo/eval.mjs) that scores hit@5 against eval/golden.yaml.

export const CANDIDATES_PER_STAGE = 8;
export const TOP_K = 5;

export const tokenize = (text) => text.toLowerCase().match(/[a-z0-9]+/g) || [];

// Okapi BM25, identical constants and formula to das/bm25.py.
export class BM25 {
  constructor(docs, k1 = 1.5, b = 0.75) {
    this.k1 = k1;
    this.b = b;
    this.docTokens = docs.map(tokenize);
    this.docLen = this.docTokens.map((t) => t.length);
    this.avgLen = docs.length
      ? this.docLen.reduce((a, x) => a + x, 0) / docs.length
      : 0;
    this.docTf = this.docTokens.map((toks) => {
      const tf = new Map();
      for (const t of toks) tf.set(t, (tf.get(t) || 0) + 1);
      return tf;
    });
    const df = new Map();
    for (const tf of this.docTf)
      for (const term of tf.keys()) df.set(term, (df.get(term) || 0) + 1);
    const n = docs.length;
    this.idf = new Map(
      [...df].map(([term, d]) => [term, Math.log(1 + (n - d + 0.5) / (d + 0.5))]),
    );
  }

  score(query, i) {
    const tf = this.docTf[i];
    const dl = this.docLen[i] || 1;
    let s = 0;
    for (const term of tokenize(query)) {
      const f = tf.get(term);
      if (!f) continue;
      s +=
        (this.idf.get(term) || 0) *
        ((f * (this.k1 + 1)) /
          (f + this.k1 * (1 - this.b + (this.b * dl) / this.avgLen)));
    }
    return s;
  }

  top(query, k) {
    return this.docTokens
      .map((_, i) => [i, this.score(query, i)])
      .filter(([, s]) => s > 0)
      .sort((a, b) => b[1] - a[1])
      .slice(0, k);
  }
}

// Keep only the best-scoring window per clip, preserving rank order.
export function dedupeByClip(ranked, windows, k) {
  const seen = new Set();
  const out = [];
  for (const [idx, score] of ranked) {
    const clip = windows[idx].clip_id;
    if (seen.has(clip)) continue;
    seen.add(clip);
    out.push([idx, score]);
    if (out.length === k) break;
  }
  return out;
}

// Cosine over normalized vectors = dot product (embeddings: Float32Array n*dim).
export function semanticTop(embeddings, dim, qvec, k) {
  const n = embeddings.length / dim;
  const sims = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    let s = 0;
    const off = i * dim;
    for (let j = 0; j < dim; j++) s += embeddings[off + j] * qvec[j];
    sims[i] = s;
  }
  const order = [...sims.keys()].sort((a, b) => sims[b] - sims[a]);
  return order.slice(0, k);
}

// embed(query) -> Float32Array(dim), normalized.
// rerank(query, texts[]) -> number[] cross-encoder scores.
export function createSearcher({ windows, records, embeddings, dim, embed, rerank }) {
  const bm25 = new BM25(windows.map((w) => w.index_text));

  return async function search(query, k = TOP_K) {
    // Stage 1: candidate union
    const qvec = await embed(query);
    const semantic = new Set(semanticTop(embeddings, dim, qvec, CANDIDATES_PER_STAGE));
    const lexical = new Set(bm25.top(query, CANDIDATES_PER_STAGE).map(([i]) => i));
    const candidates = [...new Set([...semantic, ...lexical])].sort((a, b) => a - b);
    if (!candidates.length) return [];

    // Stage 2: rerank, then keep each clip's best window
    const scores = await rerank(query, candidates.map((i) => windows[i].index_text));
    let ranked = candidates
      .map((c, j) => [c, scores[j]])
      .sort((a, b) => b[1] - a[1]);
    ranked = dedupeByClip(ranked, windows, k);

    return ranked.map(([idx, score]) => {
      const w = windows[idx];
      const rec = records[w.clip_id];
      const inSem = semantic.has(idx);
      const inLex = lexical.has(idx);
      return {
        clip_id: w.clip_id,
        title: rec.title,
        program: rec.program,
        date: rec.date,
        decade: rec.decade,
        excerpt: w.text,
        start_s: w.start,
        end_s: w.end,
        asr_confidence: w.confidence,
        rerank_score: Math.round(score * 1e4) / 1e4,
        matched_by: (inSem ? 'semantic' : '') + (inSem && inLex ? '+lexical' : '') + (!inSem && inLex ? 'lexical' : ''),
        tags: rec.tags,
        source_url: rec.source_url,
        rights: rec.rights,
        file_url: rec.file_url,
      };
    });
  };
}
